### PR TITLE
C#: Use explicit loop instead of Sequence.Create when possible

### PR DESCRIPTION
**Pull request by lukemaurer**
There's overhead in Sequence.Create having to call the passed-in function for each element.  If we know the function is a lambda, we can instead put the body of the lambda in a loop that populates the sequence.

### CODE REVIEW

**Comment by lukemaurer** [comment]
Microbenchmarked using a simple one-liner:

```
method Main() {
  print |seq(100000000, i => i)|;
}
```

Average wall time for three runs (after one cold run) was 1964ms on master and 1763ms on branch.

**Comment by seanmcl** [comment]
@RustanLeino How much are we willing to sacrifice succinct, clear code for inlined or otherwise less nice code for the sake of performance?

File: Source/Dafny/Compiler-Csharp.cs [hunk 1]
```diff
@@ -1709,11 +1721,41 @@ protected class ClassWriter : IClassWriter {
     }
 
     protected override void EmitSeqConstructionExpr(SeqConstructionExpr expr, bool inLetExprBody, TargetWriter wr) {
-      wr.Write("{0}<{1}>.Create(", DafnySeqClass, TypeName(expr.Type.AsSeqType.Arg, wr, expr.tok));
-      TrExpr(expr.N, wr, inLetExprBody);
-      wr.Write(", ");
-      TrExpr(expr.Initializer, wr, inLetExprBody);
-      wr.Write(")");
+      var elmtType = expr.Type.AsSeqType.Arg;
```
Comment (seanmcl) [inline Source/Dafny/Compiler-Csharp.cs]: Could we write a private method with the `LambdaExpr` code with a comment saying that this is equivalent but faster than the corresponding generic code, and then call that here?

File: Source/Dafny/Compiler-Csharp.cs [hunk 2]
```diff
@@ -1709,11 +1721,41 @@ protected class ClassWriter : IClassWriter {
...
+        DeclareLocalVar(lengthVar, indexType, expr.tok, expr.N, inLetExprBody, wrLamBody);
+        var arrVar = FreshId("arr");
+        wrLamBody.Write("var {0} = ", arrVar);
+        var wrDims = EmitNewArray(elmtType, expr.tok, dimCount: 1, mustInitialize: false, wr: wrLamBody);
+        Contract.Assert(wrDims.Count == 1);
+        wrDims[0].Write(lengthVar);
+        wrLamBody.WriteLine(";");
+
+        var intIxVar = FreshId("i");
+        var wrLoopBody = wrLamBody.NewBlockWithPrefix("for (int {0} = 0; {0} < {1}; {0}++)", "", intIxVar, lengthVar);
```
Comment (RustanLeino) [inline Source/Dafny/Compiler-Csharp.cs]: `lengthVar` is of type `indexType`, which is probably `BigInteger`. If so, the comparison `{0} < {1}` will compare an `int` with ` BigInteger`. This is okay, but could be slightly improved by first sticking `lengthVar` into a `uint`.

Comment (lukemaurer) [reply Source/Dafny/Compiler-Csharp.cs]: For that matter, should `i` be a `long`?  There are other ways we support (in C#) sequences of 64-bit lengths.

File: Source/Dafny/Compiler-Csharp.cs (hunk 1 above)
Comment (lukemaurer) [reply Source/Dafny/Compiler-Csharp.cs]: Sure, done.

**Pull request merged**